brickschema/validate.py

```diff
@@ -0,0 +1,423 @@
+                g.add((focusNode, resultPath, valueNode))
+                violation.add((BNode(), BSH['offendingTriple'], g))
+                return
+            else:
+                # Without valueNode, we look for constraint, such as
+                # sh:class <class> and sh:minCount <number>
+                cComp = self.__violationPredicateObj(violation,
+                                                     'sh:sourceConstraintComponent')
+                c = cComp.split('#')[1].replace('ConstraintComponent', '')
+                cPred = 'sh:' + c[0].lower() + c[1:]
+                cObj = self.__prefix(self.__violationPredicateObj(violation, cPred))
+
+                g = Graph()
+                g.add((focusNode, resultPath, Literal(f'{cPred} {cObj}')))
+                violation.add((BNode(), BSH['offenderHint'], g))
+
+            return
+        # end of if resultPath:
+
+        # Without sh:resultPath or sh:value in the violation. We are currently only
+        # concerned with the RDFS.domain shape.
+        sourceShape = self.__violationPredicateObj(violation, 'sh:sourceShape')
+        if sourceShape.endswith('DomainShape'):
+            (bsh, shapeName) = sourceShape.split('#')
+
+            # For a brick property xyz with RDFS.domain predicate, the shape's name
+            # is bsh:xyzDomainShape.  Here we tease out brick:xyz to make the query.
+            brickProp = shapeName[:-len('DomainShape')]
+            path = 'brick:' + brickProp
+            fullPath = self.namespaceDict['brick'] + brickProp
+
+            # The full name (http...) of the focusNode doesn't seem to work
+            # in the query.  Therefore make a prefixed version for the query.
+            focusNode = self.__violationPredicateObj(violation, 'sh:focusNode')
+            res = self.__queryDataGraph(self.__prefix(focusNode), path, None)
+
+            # Due to inherent ambiguity of this kind of shape,
+            # multiple triples may be found.
+            for (s, p, o) in res:
+                g = Graph()
+                g.add((focusNode, URIRef(fullPath), o))
+                violation.add((BNode(), BSH['offendingTriple'], g))
+            return
+        # end of if sourceShape.endswith('DomainShape'):
+
+        # When control reaches here, a handler is missing for the violation.
+
+        self.log.error('no triple finder for violation %s' %
+                       violation.serialize(format='ttl').decode('utf-8'))
+        return
+    # end of triplesForOneViolation()
+
+# end of class Validate()
+
+class ResultsSerialize():
+    """
+    Serializes violations with extra offender info.
+    """
+
+    def __init__(self, violationList, namespaceDict, output):
+        self.log = logging.getLogger()
+        self.log.setLevel(logging.DEBUG if hasattr(sys, '_called_from_test') else logging.WARNING)
+
+        self.violationList = violationList
+        self.outFile = output
+        self.namespaceDict = namespaceDict
```

Comment:
Rather than `ResultsSerialize` being a class, can this logic be attached to the `Result` object I proposed above? that way it is automatically folded into how the results will be consumed and returned

Reply:
Agreed.